### PR TITLE
Added video quality feature

### LINKED ISSUE #41: ✨ Feature parity: Ability to change video quality
## How does this feature work on Reddit?

I want to be able to change the video quality just like on reddit, it helps saving data also it would help if we could save the preferred video quality,
## Describe how this could be implemented into Redlib

I don't have any idea of how it can be implemented. If you give me some idea of how it can be implemented, I am willing to submit a PR myself.

### PR DESCRIPTION
**Pull request by Carbrex**
May close #41 
I have tried adding the quality change feature, it works but with a delay as it waits for the current segment to finish playing. Please check it out and suggest changes.
I also found some repositories related to this. https://github.com/sahilkashyap64/hls and https://github.com/kmoskwiak/videojs-resolution-switcher but they require some more dependencies.

### CODE REVIEW

File: static/playHLSVideo.js [hunk 1]
```diff
@@ -30,12 +30,22 @@
 
             function initializeHls() {
                 newVideo.removeEventListener('play', initializeHls);
-
                 var hls = new Hls({ autoStartLoad: false });
                 hls.loadSource(playlist);
                 hls.attachMedia(newVideo);
+                console.log(hls);
```
Comment (sigaloid) [inline static/playHLSVideo.js:36]:
```suggestion
```

File: static/playHLSVideo.js [hunk 2]
```diff
@@ -61,6 +71,24 @@
                 });
             }
 
+            function addQualitySelector(videoElement, hlsInstance, availableLevels) {
+                var qualitySelector = document.createElement('select');
+                availableLevels.forEach(function (level, index) {
+                    var option = document.createElement('option');
+                    option.value = index.toString();
+                    option.text = level.height + 'p';
```
Comment (sigaloid) [inline static/playHLSVideo.js:79]:
```suggestion
                    var bitrate = (level.bitrate / 1_000_000).toFixed(2);
                    option.text = level.height + 'p (' + bitrate + ' mbps)';
```

Comment (sigaloid) [reply static/playHLSVideo.js:79]: Optionally, divide by 1000 and drop the round (leave it as an integer - kbps is specific enough).

image

I think this is the better way to go now tbh - Reddit brutally compresses video and mbps is far less useful IMO

**Comment by sigaloid** [comment]
This looks great so far! A couple of suggestions with the bitrate there. That being said, I think long-term it would be better to switch to something like dash.js - this is a good stopgap but if you're proficient with JS and want to try your hand, could be a good thing to work on. I'll open an issue for that but we can go forward with this for now

**Comment by Carbrex** [comment]
How do I add it into settings to set a default preference for video quality?

File: static/playHLSVideo.js [hunk 3]
```diff
@@ -30,12 +30,21 @@
 
             function initializeHls() {
                 newVideo.removeEventListener('play', initializeHls);
-
                 var hls = new Hls({ autoStartLoad: false });
                 hls.loadSource(playlist);
                 hls.attachMedia(newVideo);
                 hls.on(Hls.Events.MANIFEST_PARSED, function () {
-                    hls.loadLevel = hls.levels.length - 1;
+                    hls.loadLevel = 0;
```
Comment (sigaloid) [inline static/playHLSVideo.js:37]: I assume it is something to do with this - this code sets it to the lowest (it's sorted by bitrate by Reddit) but it used to be the highest.

Comment (Carbrex) [reply static/playHLSVideo.js:37]: I know that this sets it to the lowest. I wanted to ask how/where do I add some preference button in settings that lets you set some preference.

Comment (sigaloid) [reply static/playHLSVideo.js:37]: Ah, I see. I would just select the highest by default as that was the previous behavior. Now there's the chance to reduce that at least. But I don't think we really need to configure this unless there's a ton of demand for it (especially given there's so many different resolutions and bandwidths for videos Reddit returns)

Comment (Carbrex) [reply static/playHLSVideo.js:37]: Ok np. But still for my knowledge can you tell me how can I get any preference value say `post_sort` in `playHlsVideo.js`.

Comment (Carbrex) [reply static/playHLSVideo.js:37]: Also should I revert to highest?

Comment (sigaloid) [reply static/playHLSVideo.js:37]: I think that would be a good idea just to retain previous behavior, yeah.

**Comment by sigaloid** [comment]
Added some slight styling: 
image
Placed it on the right as that's typically where video quality options are. Anyway, other than this, I think it's ready to go in

**Pull request merged**